drivers/serial/uart_bcm2711.c

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright (c) 2023 honglin leng <a909204013@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT brcm_bcm2711_aux_uart
+
+/**
+ * @brief BCM2711 Miniuart Serial Driver
+ *
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <stdbool.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/irq.h>
+
+#define BCM2711_MU_IO			0x00
+#define BCM2711_MU_IER			0x04
+#define BCM2711_MU_IIR			0x08
+#define BCM2711_MU_LCR			0x0c
+#define BCM2711_MU_MCR			0x10
+#define BCM2711_MU_LSR			0x14
+#define BCM2711_MU_MSR			0x18
+#define BCM2711_MU_SCRATCH		0x1c
+#define BCM2711_MU_CNTL			0x20
+#define BCM2711_MU_STAT			0x24
+#define BCM2711_MU_BAUD			0x28
+
+#define BCM2711_MU_IER_TX_INTERRUPT	(1 << 1)
+#define BCM2711_MU_IER_RX_INTERRUPT	(1 << 0)
+
+#define BCM2711_MU_IIR_RX_INTERRUPT	(1 << 2)
+#define BCM2711_MU_IIR_TX_INTERRUPT	(1 << 1)
+#define BCM2711_MU_IIR_FLUSH		0xc6
+
+#define BCM2711_MU_LCR_7BIT		2
+#define BCM2711_MU_LCR_8BIT		3
+
+#define BCM2711_MU_LSR_TX_IDLE		(1 << 6)
+#define BCM2711_MU_LSR_TX_EMPTY		(1 << 5)
+#define BCM2711_MU_LSR_RX_OVERRUN	(1 << 1)
+#define BCM2711_MU_LSR_RX_READY		(1 << 0)
+
+#define BCM2711_MU_CNTL_RX_ENABLE	(1 << 0)
+#define BCM2711_MU_CNTL_TX_ENABLE	(1 << 1)
+
+typedef unsigned long mem_addr_t;
+
+struct bcm2711_uart_config {
+	DEVICE_MMIO_ROM; /* Must be first */
+	mem_addr_t base;
+	uint32_t baud_rate;
+	uint32_t clocks;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct bcm2711_uart_data {
+	DEVICE_MMIO_RAM; /* Must be first */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+bool bcm2711_mu_lowlevel_can_getc(mem_addr_t base)
+{
+	return (!(sys_read32((base + BCM2711_MU_LSR)) & BCM2711_MU_LSR_RX_READY))
+		? false : true;
+}
+
+uint8_t bcm2711_mu_lowlevel_getc(mem_addr_t base)
+{
+	/* Wait until there is no data in the FIFO */
+	while (!bcm2711_mu_lowlevel_can_getc(base))
+		;
+
+	return (char)(sys_read32((base + BCM2711_MU_IO)) & 0xFF);
+}
+
+bool bcm2711_mu_lowlevel_can_putc(mem_addr_t base)
+{
+	mem_addr_t lsr = (base + BCM2711_MU_LSR);
+
+	if (!(sys_read32(lsr) & BCM2711_MU_LSR_TX_EMPTY))
+		return false;
+
+	return true;
+}
+
+void bcm2711_mu_lowlevel_putc(mem_addr_t base, uint8_t ch)
+{
+	/* Wait until there is data in the FIFO */
+	while (!bcm2711_mu_lowlevel_can_putc(base))
+		;
+
+	/* Send the character */
+	sys_write32(ch, (base + BCM2711_MU_IO));
```

Comment:
remove the unnecessary parentheses everywhere:
```suggestion
	sys_write32(ch, base + BCM2711_MU_IO);
```